### PR TITLE
New demos2

### PR DESCRIPTION
**Pull request by dlyr**
* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
add more texture generation to mesh/draw primitives

* **What is the current behavior?** (You can also link to an open issue here)
missing tex coord generations for most primitives

* **What is the new behavior (if this is a feature change)?**
add tex coord generation for (and also todo)
- [x] parametric sphere
- [x] sharp box
- [x] torus
- [ ] triangles
- [ ] to be completed

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
old client code should work without change.

* **Other information**:
TODO:
- [x] add documentation

### CODE REVIEW

**Comment by dlyr** [comment]
Preview of new DrawPrimitivesDemo

**Comment by hiergaut** [comment]
It's ok for the merged of torus in your PR.
Be careful, there will be a conflict with this PR during the merge on release candidate due to the PR #900 that has been introduced in the meantime. Of the non-use of standard attributes (hard) in the code.

**Review by nmellado — APPROVED** [review]
Seems fine to me

**Comment by dlyr** [comment]
Well there is still some todos ... At least add some doc

File: src/Core/Geometry/MeshPrimitives.inl [hunk 1]
```diff
@@ -1,98 +1,202 @@
...
+    WAI whColor =
+        color ? topoMesh.addWedgeAttrib<Vector4>( getAttribName( MeshAttrib::VERTEX_COLOR ) )
+              : WAI::Invalid();
+    WAI whTexCoord =
+        gtc ? topoMesh.addWedgeAttrib<Vector3>( getAttribName( MeshAttrib::VERTEX_TEXCOORD ) )
+            : WAI::Invalid();
+
+    TopologicalMesh::VertexHandle vhandles[( stacks - 1 ) * slices + 2];
+    Vector3Array topoTexCoords;
+    topoTexCoords.reserve( ( stacks - 1 ) * slices + 2 + 2 * slices );
```
Comment (MathiasPaulin) [inline src/Core/Geometry/MeshPrimitives.inl:33]: Why is there a different number of tex coord than number of vertices/normals ?

File: src/Engine/Data/DrawPrimitives.cpp [hunk 2]
```diff
@@ -244,7 +244,7 @@ LineMeshPtr CircleArc( const Core::Vector3& center,
 }
 
 MeshPtr Sphere( const Core::Vector3& center, Scalar radius, const Core::Utils::Color& color ) {
-    auto geom   = makeGeodesicSphere( radius, 2 );
+    auto geom   = makeGeodesicSphere( radius, 4, color );
```
Comment (MathiasPaulin) [inline src/Engine/Data/DrawPrimitives.cpp:247]: why "4". Should'nt it be a typed constant or a function parameter?

File: src/Engine/Data/DrawPrimitives.cpp [hunk 3]
```diff
@@ -253,9 +253,21 @@ MeshPtr Sphere( const Core::Vector3& center, Scalar radius, const Core::Utils::C
...
-        Ra::Core::Geometry::getAttribName( Ra::Core::Geometry::MeshAttrib::VERTEX_COLOR ),
-        Core::Vector4Array { geom.vertices().size(), color } );
+    return make_shared<Mesh>( "Sphere Primitive", std::move( geom ) );
+}
+
+MeshPtr ParametricSphere( const Core::Vector3& center,
+                          Scalar radius,
+                          const Core::Utils::Color& color,
+                          bool generateTexCoord ) {
+    auto geom   = makeParametricSphere<32, 32>( radius, color, generateTexCoord );
```
Comment (MathiasPaulin) [inline src/Engine/Data/DrawPrimitives.cpp:263]: I think we should avoid hard-coded litteral constants and rather uses typed constants (i.e `constexpr const int slices = 32;`)

File: src/Core/Geometry/MeshPrimitives.inl (hunk 1 above)
Comment (dlyr) [reply src/Core/Geometry/MeshPrimitives.inl:33]: for pole I think

File: src/Engine/Data/DrawPrimitives.cpp (hunk 2 above)
Comment (dlyr) [reply src/Engine/Data/DrawPrimitives.cpp:247]: why not, that's from previous code. 
Anyway I think DrawPrimitives specific code should disapear and use Core directly.
Core function already takes parameters or are templated with such constant.

File: src/Engine/Data/DrawPrimitives.cpp (hunk 3 above)
Comment (dlyr) [reply src/Engine/Data/DrawPrimitives.cpp:263]: Same as above, I think we will remove these call in a near futur. But improvement welcome if you want to fix these hard codded values.

File: src/Core/Geometry/MeshPrimitives.inl [hunk 4]
```diff
@@ -101,19 +205,25 @@ TriangleMesh makeParametricTorus( Scalar majorRadius,
...
-                                          iu * V + ( ( iv + 1 ) % V ) ) );
+            texCoords.emplace_back( iu * du, iv * dv, 0_ra );
+
+            if ( iu != U && iv != V ) {
+                indices.push_back( Vector3ui( iu * ( V + 1 ) + iv,
+                                              ( ( iu + 1 ) ) * ( V + 1 ) + iv,
+                                              iu * ( V + 1 ) + ( ( iv + 1 ) ) ) );
+                indices.push_back( Vector3ui( ( ( iu + 1 ) ) * ( V + 1 ) + iv,
+                                              ( ( iu + 1 ) ) * ( V + 1 ) + ( ( iv + 1 ) ),
+                                              iu * ( V + 1 ) + ( ( iv + 1 ) ) ) );
```
Comment (hiergaut) [inline src/Core/Geometry/MeshPrimitives.inl:216]:
```suggestion
                                              ( iu + 1 ) * ( V + 1 ) + iv,
                                              iu * ( V + 1 ) + ( iv + 1 ) ) );
                indices.push_back( Vector3ui( ( iu + 1 ) * ( V + 1 ) + iv,
                                              ( iu + 1 ) * ( V + 1 ) + ( iv + 1 ),
                                              iu * ( V + 1 ) + ( iv + 1 ) ) );
```

Comment (hiergaut) [reply src/Core/Geometry/MeshPrimitives.inl:216]: Oups I forgot to remove double parentheses

File: src/Core/Geometry/MeshPrimitives.inl (hunk 1 above)
Comment (dlyr) [reply src/Core/Geometry/MeshPrimitives.inl:33]: thinking again, it's for seams where u=1 meets u=0

**Review by nmellado — APPROVED** [review]

**Pull request merged**